api.go

```diff
@@ -404,59 +402,54 @@ const (
 
 // IsOnline returns true if healthcheck enabled and client is online
 func (c *Client) IsOnline() bool {
-	switch atomic.LoadInt32(&c.healthCheck) {
-	case online, unknown:
-		return true
-	}
-	return false
+	return !c.IsOffline()
+}
+
+// sets online healthStatus to offline
+func (c *Client) markOffline() {
+	atomic.CompareAndSwapInt32(&c.healthStatus, online, offline)
 }
 
 // IsOffline returns true if healthcheck enabled and client is offline
 func (c *Client) IsOffline() bool {
-	return !c.IsOnline()
+	return atomic.LoadInt32(&c.healthStatus) == offline
 }
 
 // HealthCheck starts a healthcheck to see if endpoint is up. Returns a context cancellation function
 // and and error if health check is already started
 func (c *Client) HealthCheck(hcDuration time.Duration) (context.CancelFunc, error) {
-	if atomic.LoadInt32(&c.healthCheck) == online {
-		return nil, fmt.Errorf("health check running already")
+	if atomic.LoadInt32(&c.healthStatus) == online {
+		return nil, fmt.Errorf("health check is running")
 	}
 	if hcDuration < 1*time.Second {
 		return nil, fmt.Errorf("health check duration should be atleast 1 second")
 	}
 	ctx, cancelFn := context.WithCancel(context.Background())
-	c.healthCheckCh = make(chan struct{})
-	atomic.StoreInt32(&c.healthCheck, online)
+	atomic.StoreInt32(&c.healthStatus, online)
 	probeBucketName := randString(60, rand.NewSource(time.Now().UnixNano()), "probe-health-")
 	go func(duration time.Duration) {
 		timer := time.NewTimer(duration)
 		defer timer.Stop()
 		for {
 			select {
 			case <-ctx.Done():
-				close(c.healthCheckCh)
-				atomic.StoreInt32(&c.healthCheck, unknown)
+				atomic.StoreInt32(&c.healthStatus, unknown)
 				return
 			case <-timer.C:
-
 				timer.Reset(duration)
 				// Do health check the first time and ONLY if the connection is marked offline
-				if c.IsOffline() || c.lastOnline.IsZero() {
+				if c.IsOffline() {
 					_, err := c.getBucketLocation(context.Background(), probeBucketName)
 					if err != nil && IsNetworkOrHostDown(err, false) {
-						atomic.CompareAndSwapInt32(&c.healthCheck, online, offline)
-					} else {
-						switch ToErrorResponse(err).Code {
-						case "NoSuchBucket", "AccessDenied", "":
-							c.lastOnline = time.Now()
-							atomic.CompareAndSwapInt32(&c.healthCheck, offline, online)
-						}
+						c.markOffline()
```

Comment:
since this is already offline, no need to mark offline again

Reply:
> since this is already offline, no need to mark offline again

yeah but I think if IsNetworkOrHostDown() doesn't return true, then we should switch online in all cases